here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/PgWriter.kt

```diff
@@ -186,13 +187,28 @@ class PgWriter(
                         CreateCollection(session).execute(mapOf(write), write)
                     )
 
-                    WriteOp.UPSERT -> cachedTupleNumber(
-                        write,
-                        upsertCollection(mapOf(write), write)
-                    )
+                    WriteOp.UPSERT ->
+                        if (write.id == null || previousMetadataProvider.get(
+                                VIRT_COLLECTIONS,
+                                write.id!!
+                            ) == null )
+                        {
```

Comment:
We should not workaround the compiler, you can change the code into:
```kotlin
val id = write.id
if (id == null || previousMetadataProvider.get(VIRT_COLLECTIONS,id) == null ) {
  ...
```
Apart, this looks unhealthy, actually, what happens if the `write.id` is _null_ and/or the _previousMetadataProvider.get_ returns _null_?
When I understand this right, you should move this code out, and handle it like following:
- UPSERT without "id" should throw an NakshaException, because that is not possible
- UPSERT with no previous state (_previousMetadataProvider.get_ returns _null_), should cause an create, while when there is a previous state, it should be an UPDATE. In other words, then `when` should not have an UPSERT case, rather the `write.op` should be updated to either CREATE or DELETE before the when, if we find an UPSERT request (IMHO).